### ISSUE TITLE
customizing number of classes

### ISSUE THREAD

**Issue by Shaunlipy** [ayooshkathuria/YOLO_v3_tutorial_from_scratch#6]
Hi, I'm wondering how would I be able to modify it to train on my own 10-class dataset?

I tried changing the cfg file 3 'Yolo' sections' classes to 10. But I guess I need to modify the darknet forward pass somewhere. 

thanks

**Comment by ayooshkathuria** [comment]
I'm sorry, but the training module isn't ready as of yet, and you won't be able to train with your classes as of now. It'll take me a couple of weeks to do that since I have my undergraduate thesis presentation in first week of May. 

The number of classes is controlled by the convolutional layer right behind YOLO layer which you will see always has 255 filters. Why is that? 3 bounding boxes per cell. For each bounding box, we have 5 attributes and 80 class scores for COCO. That makes 3 * (5 + 80) = 255. 

If you want to train on your own data, you need to load darknet-53 weights file from official YOLO website, then strip the convolutional layer having 255 filters just before YOLO, you should replace it with the same convolutional layer with (3 * (5 + **10**))  = 45 filters. 

Then you'll have to figure out how to annotate your bounding boxes, write the loss function and call `backward` function on your graph in PyTorch. I'm interning at a place that requires me to train on a custom dataset as well, and when I do, I'll write another tutorial, with details on visualizing and hyperparameter tuning. 

But first 10 days of may, I'm away to my college for presenting my undergrad thesis.  So, training module shouldn't be possible for at least a couple of weeks from my side.

**Comment by Shaunlipy** [comment]
Hi, thanks for such quick reply. Yeah, I did have a general understanding of your code. (btw, appreciate for your work)~!

For your "def create_modules" function on line 163:
you always have "output_filters.append(filters)". I'm not sure if it is save to change the filters variable to 45 (3* (5 + 10) in 10 classes).
The pretrained weights is not a top concern right now.
I'll try it, and if works, can share some thoughts with you.

**Comment by ayooshkathuria** [comment]
The line 163 does nothing but append the number of feature maps of every layer to the list `output_filters` so we can make a calculation for the depth of feature map outputted by a `Route` layer later onwards. 

It has nothing to do with 45 filters. You just have to change the values in config file, then load the darknet-53 file, which is not the same as yolov3.weights file (former is for training, while other is for inference out of the box). Correct me if I'm wrong anywhere?

**Comment by Shaunlipy** [comment]
Yeah, it makes sense.

One more question:

Is the code on utils.py line 105: prediction[:,:,:4] = box_corner[:,:,:4]
supposed to be the other way around?

why the prediction's object score gets re-assigned?

Thanks

**Comment by Shaunlipy** [comment]
You are right, my bad. Didn't look carefully
